### ISSUE TITLE
barcheck fail: note crosses barline when repeating

### ISSUE THREAD

**Issue by suntong** [ssb22/jianpu-ly#45]
Please take a look

```
4/4,4

KeepLength

R{
1' 5. q3 c3 1   2. q1 c2 3

} A{
1. q1 c6, 5, 5, - - %|
%1. q1 c6, 5, 5, - -
}
```

It works, but when uncommenting those "`%`" then I'll get 

`Error: (notesHad=1 5 3 3 1 2 1 2 3 1 1 6 5 5 - - 1) barcheck fail: note crosses barline at "1" with 0 beams (24 skipped from 48 to 72, bypassing 64), scoreNo=1 barNo=5 (but the error could be earlier)`

This looks like a bug to me, or ...? Thanks.

**Comment by ssb22** [comment]
Yes it's a bug, it's not resetting `notehead_markup.barPos` in the `repeatStack and repeatStack[-1][0]` part, which means the bar-check gets thrown off if you end the first-time on an incomplete bar (which you're doing here because of a pickup).  I'll try to fix this later, meanwhile you can just comment out the line of code with `barcheck fail` in it and it should work (although if you then run that version on input with any *real* barcheck errors, it won't pick them up and you'll get Lilypond warnings and messed-up output)

**Comment by suntong** [comment]
Hmm... Having disabled this line

https://github.com/ssb22/jianpu-ly/blob/696c9f5e6b21ff8c3b42cc12ad44d0fcf9780545/jianpu-ly.py#L527

I now get:

`Error: Incomplete bar at end of score 1 (pos 32)`

So I think I'll wait for your official fix.

**Comment by suntong** [comment]
yes, problem fixed! Hope #46 can be fixed soon as well. thx.

**Comment by suntong** [comment]
Re-use this issue so that all the same symptom will have a single place to discuss. 

Take a look at this code:

```
SeparateTimesig 1=C 4/4

KeepLength

2\\ 3 56  3\\. 2\\\ 3\\ 2 1  c1 - |
%2\\ 3 5 6  3\\. 2\\\ 3\\ 2 1  c1 - |
```

The first line, when `56` are together, gives wrong outcome:

> ![Image](https://github.com/user-attachments/assets/5686f706-8c89-4d33-824c-178e78771a8f)

But if I separate them, ie. un-comment the second line, I'll get the error of:

`Error: (notesHad=2 3 5 6 3 2 3 2 1 1 -) barcheck fail: note crosses barline at "-" with 0 beams (16 skipped from 52 to 68, bypassing 64), scoreNo=1 barNo=1 (but the error could be earlier)`

Why is that and how to fix pls?

**Comment by ssb22** [comment]
This this is correct because in 4/4 time you can only have 4 crotchets which is 8 quavers or 16 semiquavers.  And as the second half of that bar is already taken up by 2 crotchets, we only have room for 8 semiquavers and you're trying to put 9 in.

If the original manuscript has 9 notes in the space of 8 then you need to tell it this explicitly by using a "tuplet": `KeepLength 9[ 2\\ 3 5 6 3\\. 2\\\ 3\\ 2 1 ] c1 - |`  the extra `9[` ... `]` tells the program that this instance of 9 notes in the space of 8 is not an error and will speed up the MIDI playing accordingly.

The manuscript you're working from should have a little "9" printed near these notes too, but some publishers do get lazy and leave it out.  Unfortunately automatically detecting tuplets can be difficult, and it's usually a typo so the code is probably right to say "error" unless you explicitly tell it it's a tuplet.

**Comment by suntong** [comment]
Oh!!! duh! I cannot even count! 
So sorry to bother you with such silly mistakes. 
thanks
